routing/cross_mwm_index_graph.hpp

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include "routing/segment.hpp"
+
+#include <vector>
+
+namespace routing
+{
+/// \brief This is an interface for cross mwm routing.
+// @TODO(bykoianko) This interface will have two implementations.
+// * For orsm cross mwm section.
+// * For A* cross mwm section
+class CrossMwmIndexGraph
+{
+public:
+  /// \brief Transition segment is a segment which is crossed by mwm border. That means
+  /// start and finsh of such segment has to lie at different mwms. If a segment is
+  /// crossed by mwm border but its start and its finish lie in the same mwm it's not
+  /// a transition segment.
+  /// For most cases there is only one transition segment for a transition feature.
+  /// Transition segment at the picture below is marked as *===>*.
+  /// Transition feature is a feature which contains one or more transition segments.
+  /// Transition features at the picture below is marked as *===>*------>*.
+  /// Every transition feature is duplicated in two neibouring mwms.
+  /// The duplicate is called "twin" at the picture below.
+  /// For most cases a transition feature contains only one transition segment.
+  /// -------MWM0---------------MWM1----------------MWM2-------------
+  /// |                   |                 |                       |
+  /// | F0 in MWM0   *===>*------>*    *===>*------>*  F2 of MWM1   |
+  /// | Twin in MWM1 *===>*------>*    *===>*------>*  Twin in MWM2 |
+  /// |                   |                 |                       |
+  /// | F1 in MWM0      *===>*---->*      *===>*--->*  F3 of MWM1   |
+  /// | Twin in MWM1    *===>*---->*      *===>*--->*  Twin in MWM2 |
+  /// |                   |                 |                       |
+  /// ---------------------------------------------------------------
+  /// There're two kinds of transition segments:
+  /// * enter transition segments are segments which are enters to theirs mwm;
+  /// * exit transition segments are segments which are exits from theirs mwm;
+  /// \returns true if |s| is a exit (|isOutgoing| == true) or enter (|isOutgoing| == false)
+  /// transition segment.
+  virtual bool IsTransition(Segment const & s, bool isOutgoing) const = 0;
+
+  /// \brief Fills |twin| with a dublicate of |s| transition segment in neibouring mwm.
+  /// If |s| an enter transition segment fills |twin| with an appropriate exit transition segment.
+  /// If |s| an exit transition segment fills |opposite| with an appropriate enter transition segment.
+  /// \note GetTwin(...) shall be called only if IsTransition(s, ...) returns true.
+  virtual void GetTwin(Segment const & s, Segment & twin) const = 0;
```

Comment:
I'm not sure. As we agreed while talking to be on the safe side I'm changing output param to vector<Segment> & twins. If there's only one twin for every transition segment vector<Segment> & twins will be changed to Segment & twin